cql3/statements/create_view_statement.cc

```diff
@@ -164,6 +164,15 @@ std::pair<view_ptr, cql3::cql_warnings_vec> create_view_statement::prepare_view(
                 "before being replayed.", column_family(), _base_name.get_column_family()));
     }
 
+    if (_properties.properties()->get_gc_grace_seconds() == 0) {
+        throw exceptions::invalid_request_exception(fmt::format(
+                "Cannot create materialized view '{}' with gc_grace_seconds "
+                "of 0 for base table '{}' , since this value is "
+                "used to TTL undelivered updates. Setting gc_grace_seconds "
+                "too low might cause undelivered updates to expire "
+                "before being replayed.", column_family(), _base_name.get_column_family()));
```

Comment:
> Note about the commit message: you say "The view update is stored in a batch log for better durability guarantees". This doesn't actually happen.

It doesn't happen in Scylla, but it does happen in Cassandra. I was basing my comment on what I understood from http://www.doanduyhai.com/blog/?p=1930. Perhaps I should have been more clear that I'm speaking about Cassandra there, but I thought the first paragraph makes it clear.

> 
> Maybe you're talking about _hints_, stored only when a remote node is down?
> 
> But in that case, what actually gets GC'ed just the tombstones? If we expire tombstones that we store in hints, it sounds like a bug in hints, not something we should hide in the MV code :-(

Unsent hints are garbage-collected after gc_grace_seconds, regardless whether they are tombstones or not - that's just a part of their design. I think this behavior makes sense for the general use case at least - without the GC behavior the following case can happen:

- Hint to partition P is stored on node A,
- Partition P is deleted on node B by a tombstone with higher timestamp than the hint,
- gc_grace_seconds pass, the tombstone on node B gets garbage collected,
- Hint gets replayed by A and sent to B - this leads to data resurrection.

Cassandra batchlog behaves similarly.

Reply:
It did confuse me - I didn't understand that much of the text in the commit message is actually about Cassandra, not Scylla. I'm not sure, by the way, why we need to explain here how Cassandra works, but I guess if you already researched it it can't hurt to include it - but perhaps you can just reorganize it to make it clearer where you're talking about Scylla - which has HINTS - and where you're talking about Cassandra that has other stuff.

By the way, I'm at a loss why we don't just fix the hints bug instead of working around it. Hints are like repair - you need to apply whatever comes through it - both live data and tombstones. You're not supposed to GC tombstones you get in hints before even applying them...

Reply:
> You're not supposed to GC tombstones you get in hints before even applying them...

It's not that we expire tombstones in hints, we completely drop hints that are older than gc_grace_seconds, no matter whether they hold tombstones live data.

As for whether it's a good idea - I'm not entirely sure. Hints are not guaranteed to be reliable, so dropping them after gc_grace_seconds might be a better idea than dropping them because it avoids data resurrection.

Reply:
Well, maybe it's a bug to "completely drop hints that are older than gc_grace_seconds"? How the @#!? is gc_grace_seconds related to dropping old hints? If I set gc_grace_seconds to 1 second (because you don't allow me to set it to 0), how will I not have problems now that all hints supposedly get discarded after 1 second?
I understand that we might have various rules for dropping hints, but I don't see why they should be paramterized on gc_grace_seconds, which has nothing to do with hints.
Is this just a bug? Shouldn't we fix it instead of this patch trying to paper over this bug in a really strange way?

Reply:
The gc grace period model assumed that repair will run in the grace period, and repair is supposed to flush hints, right?
And the hints are not needed after repair.
The bottom line is that repair should be the mechanism to "trim" hints, not a time-based mechanism.
That's why tombstone_gc_mode=repair is so important.

That said, we also have the odd cases of tombstone_gc_mode=immediate or timeout mode and gc_grace_seconds=0, where we can't expect repair to run before garbage collection.  That may be suitable when using rf=1.

So what about the hints in this case?
Is this a problem only for views?
Why isn't it a problem for regular tables as well?
Maybe we should disable hints (and view hint) for tables using immediate tombstone_gc alltogether?